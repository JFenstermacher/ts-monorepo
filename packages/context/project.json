{
  "name": "context",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/context/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/context",
        "main": "packages/context/src/index.ts",
        "tsConfig": "packages/context/tsconfig.lib.json",
        "assets": ["packages/context/*.md"]
      }
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs context {args.ver} {args.tag}"
      },
      "dependsOn": ["build"]
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/context/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "packages/context/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  },
  "tags": []
}
