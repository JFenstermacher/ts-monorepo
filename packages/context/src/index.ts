export * from './lib/context';
export * from './lib/types';
