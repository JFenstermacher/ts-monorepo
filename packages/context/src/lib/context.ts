import {
  ContextDefaults,
  NameAttributeKey,
  Environment,
  Tags,
  StringCase,
  ContextInput,
  ContextType,
  ContextIdAttributes,
  NameAttributeKeys
} from './types';
import { applyStrCase } from './utils';

const CONTEXT_DEFAULTS: ContextDefaults = {
  enabled: true,
  organization: "JFenstermacher",
  labelOrder: ["service", "environment", "name"],
  delimiter: "-",
  idLengthLimit: 255,
  tags: {},
  labelKeyCase: "lowercase",
  labelValueCase: "lowercase",
  labelsAsTags: ["project", "service", "environment", "name"]
}

const ID_KEYS: NameAttributeKey[] = [
  "organization",
  "project",
  "service",
  "environment",
  "name",
  "attributes"
];


export default class Context {
  organization?: string
  project: string
  service: string
  environment: Environment
  name?: string
  attributes?: string[]

  enabled: boolean
  labelOrder: NameAttributeKeys
  delimiter: string
  idLengthLimit: number
  tags: Tags
  labelKeyCase: StringCase
  labelValueCase: StringCase
  labelsAsTags: NameAttributeKeys

  id: string
  idFull: string

  constructor(context: ContextInput) {
    const merged = { ...CONTEXT_DEFAULTS, ...context };

    this._validate(merged);

    this.organization = merged.organization;
    this.project = merged.project;
    this.service = merged.service;
    this.environment = merged.environment;
    this.name = merged.name;
    this.attributes = merged.attributes;
    this.enabled = merged.enabled;
    this.labelOrder = merged.labelOrder;
    this.delimiter = merged.delimiter;
    this.idLengthLimit = merged.idLengthLimit;
    this.labelKeyCase = merged.labelKeyCase;
    this.labelValueCase = merged.labelValueCase;
    this.labelsAsTags = merged.labelsAsTags;

    this.idFull = this._constructIdFull(merged);
    this.id = this.idFull.substring(0, this.idLengthLimit);

    this.tags = this._constructTags(merged);
  }

  get context(): ContextType {
    return {
      organization: this.organization,
      project: this.project,
      service: this.service,
      environment: this.environment,
      name: this.name,
      attributes: this.attributes,
      enabled: this.enabled,
      labelOrder: this.labelOrder,
      delimiter: this.delimiter,
      idLengthLimit: this.idLengthLimit,
      labelKeyCase: this.labelKeyCase,
      labelValueCase: this.labelValueCase,
      labelsAsTags: this.labelsAsTags,
      tags: this.tags
    }
  }

  _constructIdFull(context: ContextType) {
    const idAttributes = this._getIdAttributes(context);

    return Object.values(idAttributes).join(this.delimiter);
  }

  _constructTags(context: ContextType) {
    const tags: Tags = {};

    const idAttributes = this._getIdAttributes(context);

    for (let [key, value] of Object.entries(idAttributes)) {
      if (Array.isArray(value)) {
        value = value
          .map(v => applyStrCase(v, context.labelValueCase))
          .join(",");
      }

      tags[applyStrCase(key, context.labelKeyCase)] = value;
    }

    return {
      ...tags,
      ...context.tags,
      name: this.id
    };
  }

  _getIdAttributes(context: ContextType): ContextIdAttributes {
    const idAttributes: Partial<ContextIdAttributes> = {}

    for (const label of context.labelOrder) {
      const value = context[label];

      if (value) idAttributes[label] = value as any;
    }

    return idAttributes as ContextIdAttributes;
  }


  _validate(context: ContextType) {
    if (!context.labelOrder || !context.labelOrder.length) {
      throw Error("At least one label has to be provided in 'labelOrder'")
    }

    const hasIdAttribute = ID_KEYS.reduce((retval, key) => {
      if (context[key]) retval = true;

      return retval;
    }, false)

    if (!hasIdAttribute) {
      throw Error(`At least one ID attribute must be defined: [${ID_KEYS.join(", ")}]`)
    }
  }
}
