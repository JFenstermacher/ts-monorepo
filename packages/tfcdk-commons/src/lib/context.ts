import {
  Environment,
  Tags,
  StringCase,
  ContextInput,
  ContextType,
  IdAttributeKeys,
  ExtractableContext,
  ContextKeys,
} from './types';
import { ID_KEYS, CONTEXT_DEFAULTS, CONTEXT_KEYS } from './consts';
import { applyStrCase } from './utils';


export class Context {
  organization?: string
  project?: string
  service?: string
  environment?: Environment
  name?: string
  attributes?: string[]

  enabled: boolean
  labelOrder: IdAttributeKeys
  delimiter: string
  idLengthLimit: number
  _tags: Tags
  labelKeyCase: StringCase
  labelValueCase: StringCase
  labelsAsTags: IdAttributeKeys

  id: string
  idFull: string
  labelTags: Tags

  constructor(context: ContextInput) {
    const merged = { ...CONTEXT_DEFAULTS, ...context };

    this._validate(merged);

    this.organization = merged.organization;
    this.project = merged.project;
    this.service = merged.service;
    this.environment = merged.environment;
    this.name = merged.name;
    this.attributes = merged.attributes;
    this.enabled = merged.enabled;
    this.labelOrder = merged.labelOrder;
    this.delimiter = merged.delimiter;
    this.idLengthLimit = merged.idLengthLimit;
    this.labelKeyCase = merged.labelKeyCase;
    this.labelValueCase = merged.labelValueCase;
    this.labelsAsTags = merged.labelsAsTags;
    this._tags = merged.tags

    this.idFull = this._constructIdFull(merged);
    this.id = this.idFull.substring(0, this.idLengthLimit);

    this.labelTags = this._constructLabelTags(merged);
  }

  get context(): ContextType {
    return {
      organization: this.organization,
      project: this.project,
      service: this.service,
      environment: this.environment,
      name: this.name,
      attributes: this.attributes,
      enabled: this.enabled,
      labelOrder: this.labelOrder,
      delimiter: this.delimiter,
      idLengthLimit: this.idLengthLimit,
      labelKeyCase: this.labelKeyCase,
      labelValueCase: this.labelValueCase,
      labelsAsTags: this.labelsAsTags,
      tags: this._tags
    }
  }

  get tags(): Tags {
    const formattedTags: Tags = {};

    for (const [key, value] of Object.entries(this._tags)) {
      formattedTags[applyStrCase(key, this.labelKeyCase)] = applyStrCase(value, this.labelValueCase);
    }

    return {
      ...this.labelTags,
      ...formattedTags
    }
  }

  _constructIdFull(context: ContextType) {
    const parts: string[] = [];

    for (const label of context.labelOrder) {
      const value = context[label];

      if (!value) continue;

      if (Array.isArray(value)) parts.push(...value);
      else parts.push(value);
    }

    return parts.join(context.delimiter);
  }

  _constructLabelTags(context: ContextType) {
    const labelTags: Record<string, string> = {};

    for (const label of context.labelsAsTags) {
      let value = context[label];

      if (!value) continue;

      if (!Array.isArray(value)) value = [value];

      labelTags[applyStrCase(label, context.labelKeyCase)] = value
        .map(v => applyStrCase(v, context.labelValueCase))
        .join(',');
    }

    if (context.labelsAsTags.includes("name")) labelTags[applyStrCase("name", context.labelKeyCase)] = applyStrCase(this.id, context.labelValueCase);

    return labelTags;
  }

  _validate(context: ContextType) {
    if (!context.labelOrder || !context.labelOrder.length) {
      throw Error("At least one label has to be provided in 'labelOrder'");
    }

    const existingIdAttribute = ID_KEYS.reduce((retval, key) => {
      if (context[key]) retval = true;

      return retval;
    }, false)

    if (!existingIdAttribute) {
      throw Error("At least one ID attribute has to be passed.");
    }
  }

  static extract(props: ExtractableContext, keys: ContextKeys = CONTEXT_KEYS, context: Record<string, any> = {}) {
    for (const key of keys) {
      const value = props[key];

      if (value) context[key] = value;
    }

    return new Context(context);
  }
}
