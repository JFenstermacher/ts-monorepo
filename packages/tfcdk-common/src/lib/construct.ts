import { Construct as BaseConstruct } from 'constructs';
import { Context } from './context';
import { ConstructConfig } from './types';

export class Construct extends BaseConstruct {
  context: Context

  constructor(scope: BaseConstruct, id: string, config: ConstructConfig) {
    super(scope, id);
    const context = Context.extractContext(config)

    this.context = new Context({
      ...config.context,
      ...context
    });
  }
}

