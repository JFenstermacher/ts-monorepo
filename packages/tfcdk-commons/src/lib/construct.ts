import { Construct as BaseConstruct } from 'constructs';
import { CONTEXT_KEYS } from './consts';
import { Context } from './context';
import { ConstructConfig } from './types';

export class Construct extends BaseConstruct {
  context: Context

  constructor(scope: BaseConstruct, id: string, config: ConstructConfig) {
    super(scope, id);

    this.context = this._extractContextFromConfig(config);
  }

  _extractContextFromConfig(config: ConstructConfig): Context {
    const context: Record<string, any> = config.context ?? {};

    for (const key of CONTEXT_KEYS) {
      const value = config[key];

      if (value) context[key] = value;
    }

    return new Context(context);
  }
}
