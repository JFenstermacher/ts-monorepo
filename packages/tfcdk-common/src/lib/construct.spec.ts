import { Testing } from 'cdktf';
import { Context } from './context';
import { Construct } from './construct';

describe('Construct', () => {
  it('should manage context merging and creation', () => {
    let context = new Context({
      project: "initial",
      service: "initial"
    })

    Testing.synthScope((scope) => {
      const construct = new Construct(scope, 'test', {
        project: "overtop",
        context: context
      })

      context = construct.context;
    })

    expect(context).toMatchSnapshot()
  })
});
