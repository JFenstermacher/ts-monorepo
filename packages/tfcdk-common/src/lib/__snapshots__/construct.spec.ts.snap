// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Construct should manage context merging and creation 1`] = `
Context {
  "_tags": Object {},
  "attributes": undefined,
  "delimiter": "-",
  "enabled": true,
  "environment": undefined,
  "id": "initial",
  "idFull": "initial",
  "idLengthLimit": 255,
  "labelKeyCase": "title",
  "labelOrder": Array [
    "service",
    "environment",
    "name",
  ],
  "labelTags": Object {
    "Name": "initial",
    "Project": "overtop",
    "Service": "initial",
  },
  "labelValueCase": "lowercase",
  "labelsAsTags": Array [
    "project",
    "service",
    "environment",
    "name",
  ],
  "name": undefined,
  "organization": "JFenstermacher",
  "project": "overtop",
  "service": "initial",
}
`;
